File: src/main/java/ru/ezhov/rocket/action/caching/DiskCache.java
package ru.ezhov.rocket.action.caching;

import com.google.common.hash.Hashing;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Optional;

public class DiskCache implements Cache {
    private String cacheFolder = System.getProperty("java.io.tmpdir") + File.separator;

    @Override
    public Optional<File> fromCache(URL url) {
        String sha256hexUrl = Hashing
                .sha256()
                .hashString(url.toString(), StandardCharsets.UTF_8)
                .toString();
        File file = new File(cacheFolder + "si-" + sha256hexUrl);
        if (file.exists()) {
            return Optional.of(file);
        } else {
            try (InputStream inputStream = url.openStream();
                 FileOutputStream fileOutputStream = new FileOutputStream(file)) {
                byte[] buf = new byte[256];
                int p;
                while ((p = inputStream.read(buf)) != -1) {
                    fileOutputStream.write(buf, 0, p);
                }
            } catch (Exception e) {
                e.printStackTrace();
                file = null;
            }
        }
        return Optional.ofNullable(file);
    }
}

File: cache/src/main/kotlin/ru/ezhov/rocket/action/cache/DiskCache.kt
package ru.ezhov.rocket.action.cache

import com.google.common.hash.Hashing
import java.io.File
import java.io.FileOutputStream
import java.net.URL
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.security.cert.X509Certificate
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLContext
import javax.net.ssl.SSLSession
import javax.net.ssl.X509TrustManager

class DiskCache : Cache {
    private val cacheFolder = File("./cache"        )

    override fun get(url: URL): File? {
        if (!cacheFolder.exists()) {
            cacheFolder.mkdirs()
        }
        checkQuietly()
        val sha256hexUrl = Hashing
            .sha256()
            .hashString(url.toString(), StandardCharsets.UTF_8)
            .toString()
        val file = File(cacheFolder, sha256hexUrl)
        return if (file.exists()) {
            file
        } else {
            readFromUrl(url, file)
        }
    }

    private fun readFromUrl(url: URL, file: File): File? = try {
        url.openStream().use { inputStream ->
            FileOutputStream(file).use { fileOutputStream ->
                val buf = ByteArray(256)
                var p: Int
                while (inputStream.read(buf).also { p = it } != -1) {
                    fileOutputStream.write(buf, 0, p)
                }
            }
        }
        file
    } catch (e: Exception) {
        e.printStackTrace()
        null
    }

    companion object {
        fun checkQuietly() {
            try {
                HttpsURLConnection
                    .setDefaultHostnameVerifier { _: String?, _: SSLSession? -> true }
                val context = SSLContext.getInstance("TLS")
                context.init(
                    null,
                    arrayOf<X509TrustManager>(object : X509TrustManager {
                        override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) = Unit
                        override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) = Unit
                        override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
                    }),
                    SecureRandom()
                )
                HttpsURLConnection.setDefaultSSLSocketFactory(context.socketFactory)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}